sources/academy/webscraping/scraping_basics_python/04_downloading_html.md

```diff
@@ -0,0 +1,269 @@
+
+:::
+
+Now let's test that all works. Inside the project directory create a new file called `main.py` with the following code:
+
+```py
+import httpx
+
+print("OK")
+```
+
+Running it as a Python program will verify that your setup is okay and you've installed HTTPX:
+
+```text
+$ python main.py
+OK
+```
+
+:::info Troubleshooting
+
+If you see errors or for any other reason cannot run the code above, we're sorry, but figuring out the issue is out of scope of this course.
+
+:::
+
+## Downloading product listing
+
+Now onto coding! Let's change our code so it downloads HTML of the product listing instead of printing OK. The [documentation of the HTTPX library](https://www.python-httpx.org/) provides us with examples how to use it. Inspired by those, our code will look like this:
+
+```py
+import httpx
+
+url = "https://warehouse-theme-metal.myshopify.com/collections/sales"
+response = httpx.get(url)
+
+print(response.text)
+```
+
+If you run the program now, it should print the downloaded HTML:
+
+```text
+$ python main.py
+<!doctype html>
+<html class="no-js" lang="en">
+  <head>
+    <meta charset="utf-8">
+    <meta name="viewport" content="width=device-width, initial-scale=1.0, height=device-height, minimum-scale=1.0, maximum-scale=1.0">
+    <meta name="theme-color" content="#00badb">
+    <meta name="robots" content="noindex">
+    <title>Sales</title>
+  ...
+  </body>
+</html>
+```
+
+And that's it! It's not particularly useful yet, but it's a good start of our scraper.
+
+## About HTTP
+
+Running `httpx.get(url)`, we made our first HTTP request and received our first response. HTTP is a network protocol powering most of the internet. Understanding it well is an important foundation for successful scraping, but for now it's enough to know the basic flow and terminology.
+
+HTTP is an exchange of two participants. The _client_ sends a _request_ to the _server_, which replies with a _response_. In our case, `main.py` is the client, and the technology running at `warehouse-theme-metal.myshopify.com` replies to our request as the server.
+
+<!-- TODO image basic HTTP chart -->
+
+:::tip Deep dive to HTTP
+
+The HTTP protocol is defined by several documents called RFCs, such as [RFC 7230: HTTP Message Syntax and Routing](https://www.rfc-editor.org/rfc/rfc7230) or [RFC 7231: HTTP Semantics and Content](https://www.rfc-editor.org/rfc/rfc7231). While these technical specifications are surprisingly digestible, you may also like [HTTP tutorials by MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP).
+
+:::
+
+## Checking status codes
+
+Sometimes websites return all kinds of errors. Most often because:
+
+- The server is temporarily down.
+- The server breaks under a heavy load of requests.
+- The server applies anti-scraping protections.
+- The server application is buggy and just couldn't handle our request.
```

Comment:
In 01758535dc643b129527e1086d1f709548b0f7fe I've just cut a lot off stuff from the error handling chapter. No exceptions handling, no `import sys`, just `raise_for_status()`, because in the lessons after this one that's what's in the code examples anyway.

But I didn't remove or move this part, as I felt like I need to explain why we're doing it, and that it's important to understand for successful scraping. I can imagine letting it go though, I just need to let it sink in. I'll think about it more and revisit the section. So leaving this conversation unresolved on purpose.

Reply:
I rewrote the paragraphs in b15d988.